File: Reaktoro/Core/ReactionList.py
# Reaktoro is a unified framework for modeling chemically reactive systems.
#
# Copyright © 2014-2022 Allan Leal
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library. If not, see <http://www.gnu.org/licenses/>.


from reaktoro import *
import pytest


def testReactionList():
    species = SpeciesList("H2O(aq) H+ OH- H2(aq) O2(aq) Na+ Cl- NaCl(aq) NaCl(s) O2(g)")

    db = Database(species)

    reactions = ReactionList()

    #-------------------------------------------------------------------------
    # TESTING METHOD: ReactionList::append
    #-------------------------------------------------------------------------
    reactions.append(db.reaction("H2O(aq) = H+ + OH-"))
    reactions.append(db.reaction("H2O(aq) = H2(aq) + 0.5*O2(aq)"))
    reactions.append(db.reaction("O2(g) = O2(aq)"))
    reactions.append(db.reaction("NaCl(s) = Na+ + Cl-"))
    reactions.append(db.reaction("NaCl(s)"))

    #-------------------------------------------------------------------------
    # TESTING METHOD: ReactionList::size
    #-------------------------------------------------------------------------
    assert reactions.size() == 5

    #-------------------------------------------------------------------------
    # TESTING METHOD: ReactionList::operator[](Index)
    #-------------------------------------------------------------------------
    assert reactions[0].name() == "H2O(aq) = H+ + OH-"
    assert reactions[1].name() == "H2O(aq) = H2(aq) + 0.5*O2(aq)"
    assert reactions[2].name() == "O2(g) = O2(aq)"
    assert reactions[3].name() == "NaCl(s) = Na+ + Cl-"
    assert reactions[4].name() == "NaCl(s)"

    #-------------------------------------------------------------------------
    # TESTING METHOD: ReactionList::find
    #-------------------------------------------------------------------------
    assert reactions.find("H2O(aq) = H+ + OH-")            == 0
    assert reactions.find("H2O(aq) = H2(aq) + 0.5*O2(aq)") == 1
    assert reactions.find("O2(g) = O2(aq)")                == 2
    assert reactions.find("NaCl(s) = Na+ + Cl-")           == 3
    assert reactions.find("NaCl(s)")                       == 4

    assert reactions.find("@#$") >= reactions.size()

    #-------------------------------------------------------------------------
    # TESTING METHOD: ReactionList::findWithName
    #-------------------------------------------------------------------------
    assert reactions.findWithName("H2O(aq) = H+ + OH-")            == 0
    assert reactions.findWithName("H2O(aq) = H2(aq) + 0.5*O2(aq)") == 1
    assert reactions.findWithName("O2(g) = O2(aq)")                == 2
    assert reactions.findWithName("NaCl(s) = Na+ + Cl-")           == 3
    assert reactions.findWithName("NaCl(s)")                       == 4

    assert reactions.findWithName("@#$") >= reactions.size()

    #-------------------------------------------------------------------------
    # TESTING METHOD: ReactionList::index
    #-------------------------------------------------------------------------
    assert reactions.index("H2O(aq) = H+ + OH-")            == 0
    assert reactions.index("H2O(aq) = H2(aq) + 0.5*O2(aq)") == 1
    assert reactions.index("O2(g) = O2(aq)")                == 2
    assert reactions.index("NaCl(s) = Na+ + Cl-")           == 3
    assert reactions.index("NaCl(s)")                       == 4

    with pytest.raises(Exception):
        reactions.index("@#$")

    #-------------------------------------------------------------------------
    # TESTING METHOD: ReactionList::indexWithName
    #-------------------------------------------------------------------------
    assert reactions.indexWithName("H2O(aq) = H+ + OH-")            == 0
    assert reactions.indexWithName("H2O(aq) = H2(aq) + 0.5*O2(aq)") == 1
    assert reactions.indexWithName("O2(g) = O2(aq)")                == 2
    assert reactions.indexWithName("NaCl(s) = Na+ + Cl-")           == 3
    assert reactions.indexWithName("NaCl(s)")                       == 4

    with pytest.raises(Exception):
        reactions.indexWithName("@#$")

    #-------------------------------------------------------------------------
    # TESTING METHOD: ReactionList::withNames
    #-------------------------------------------------------------------------
    filtered = reactions.withNames(["H2O(aq) = H+ + OH-", "H2O(aq) = H2(aq) + 0.5*O2(aq)"])

    assert filtered.size() == 2
    assert filtered[0].name() == "H2O(aq) = H+ + OH-"
    assert filtered[1].name() == "H2O(aq) = H2(aq) + 0.5*O2(aq)"

    with pytest.raises(Exception):
        filtered.withNames("Calcite @#$")

    #-------------------------------------------------------------------------
    # TESTING METHOD: ReactionList::__iter__
    #-------------------------------------------------------------------------
    for i, reaction in enumerate(reactions):
        assert reaction.name() == reactions[i].name()

File: Reaktoro/Core/ReactionList.cpp
// Reaktoro is a unified framework for modeling chemically reactive systems.
//
// Copyright © 2014-2024 Allan Leal
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this library. If not, see <http://www.gnu.org/licenses/>.

#include "ReactionList.hpp"

// Reaktoro includes
#include <Reaktoro/Common/Algorithms.hpp>
#include <Reaktoro/Common/Enumerate.hpp>
#include <Reaktoro/Common/Exception.hpp>
#include <Reaktoro/Common/StringList.hpp>

namespace Reaktoro {

ReactionList::ReactionList()
{}

ReactionList::ReactionList(std::initializer_list<Reaction> reactions)
: m_reactions(std::move(reactions))
{}

ReactionList::ReactionList(const Vec<Reaction>& reactions)
: m_reactions(reactions)
{}

auto ReactionList::append(const Reaction& reaction) -> void
{
    m_reactions.push_back(reaction);
}

auto ReactionList::data() const -> const Vec<Reaction>&
{
    return m_reactions;
}

auto ReactionList::empty() const -> bool
{
    return m_reactions.empty();
}

auto ReactionList::size() const -> Index
{
    return m_reactions.size();
}

auto ReactionList::operator[](Index i) const -> const Reaction&
{
    return m_reactions[i];
}

auto ReactionList::operator[](Index i) -> Reaction&
{
    return m_reactions[i];
}

auto ReactionList::find(const String& name) const -> Index
{
    return findWithName(name);
}

auto ReactionList::findWithName(const String& name) const -> Index
{
    return indexfn(m_reactions, RKT_LAMBDA(p, p.name() == name));
}

auto ReactionList::index(const String& name) const -> Index
{
    return indexWithName(name);
}

auto ReactionList::indexWithName(const String& name) const -> Index
{
    const auto idx = findWithName(name);
    error(idx >= size(), "Could not find any Reaction object with name ", name, ".");
    return idx;
}

auto ReactionList::get(const String& name) const -> const Reaction&
{
    return getWithName(name);
}

auto ReactionList::getWithName(const String& name) const -> const Reaction&
{
    return m_reactions[indexWithName(name)];
}

auto ReactionList::withNames(const StringList& names) const -> ReactionList
{
    return vectorize(names, RKT_LAMBDA(name, m_reactions[indexWithName(name)]));
}

ReactionList::operator Vec<Reaction>&()
{
    return m_reactions;
}

ReactionList::operator Vec<Reaction>const&() const
{
    return m_reactions;
}

auto operator+(const ReactionList& a, const ReactionList& b) -> ReactionList
{
    return concatenate(a, b);
}

} // namespace Reaktoro
